### PR TITLE
Add text-image search blog

### PR DESCRIPTION
**Pull request by lesters**
@kkraune @thigm85 Please review.

### CODE REVIEW

**Review by jobergum — APPROVED** [review]
Fantastic post. Unrelated to the content but @kkraune can we see if we can change the pink highlight of `stuff` ?

image

File: _posts/2021-11-28-text-image-search.md [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+smaller amounts of data.
+
+This particular sample app is based on the [CLIP model from
+OpenAI](https://openai.com/blog/clip/), which has been trained on 400 million
+`(image, text)` pairs taken from the internet. CLIP consists of two models: one
+for text and one for images. During training, images are associated with
+textual descriptions. Like with pre-trained language models, CLIP thus gains a
+basic textual understanding of image content. Actually, the language model in
+CLIP is itself a pre-trained language model for text processing, which is
+fine-tuned during this training.
```
Comment (thigm85) [inline _posts/2021-11-28-text-image-search.md:48]: AFAIK, the CLIP team decided to train their models from scratch. 

From page 4: "Due to the large size of our pre-training dataset, over-fitting
is not a major concern and the details of training CLIP are
simplified compared to the implementation of Zhang et al.
(2020). We train CLIP from scratch without initializing the
image encoder with ImageNet weights or the text encoder
with pre-trained weights."

File: _posts/2021-11-28-text-image-search.md [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+vector. Note that this means you need to pass the text query's representation
+vector; thus, it needs to be calculated outside of Vespa for this particular
+query.
+
+The text-image search sample application includes a [Python-based search
+app](https://github.com/vespa-engine/sample-apps/blob/master/text-image-search/src/python/README.md)
+which uses the CLIP model for this. The app uses
+[pyvespa](https://pyvespa.readthedocs.io/en/latest/index.html) and is
+particularly suitable for analysis and exploration. We'll take a bit closer
+look in the next section.
```
Comment (thigm85) [inline _posts/2021-11-28-text-image-search.md:241]: remove bit?

**Review by thigm85 — CHANGES_REQUESTED** [review]

File: _posts/2021-11-28-text-image-search.md (hunk 1 above)
Comment (lesters) [reply _posts/2021-11-28-text-image-search.md:48]: Good catch!

File: _posts/2021-11-28-text-image-search.md (hunk 2 above)
Comment (lesters) [reply _posts/2021-11-28-text-image-search.md:241]: :+1:

File: _posts/2021-11-28-text-image-search.md [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+
+The sample application also includes a [stand-alone Vespa
+application](https://github.com/vespa-engine/sample-apps/tree/master/text-image-search).
+Here we move the logic and CLIP model for generating the text representation
+vector into Vespa itself. This enables querying directly with text. The
+stand-alone Vespa application is more suitable for production, and we'll take a
+look later.
+
+### Pyvespa
```
Comment (kkraune) [inline _posts/2021-11-28-text-image-search.md:248]: should be consistent Pyvespa or PyVespa or pyvespa all over - I am not sure which is right ...

**Review by kkraune — APPROVED** [review]
This is great, just minor commments

**Comment by kkraune** [comment]
> Fantastic post. Unrelated to the content but @kkraune can we see if we can change the pink highlight of `stuff` ?
> 
> image

Agreed, this is is the out-of-the-box stylesheet, will look at alternatives (that will apply to all posts and - don't hold this one for this change)

File: _posts/2021-11-28-text-image-search.md [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+from PIL import Image
+
+model, preprocess = clip.load("ViT-B/32", device="cpu")
+image = preprocess(Image.open("image.png")).unsqueeze(0)
+text = clip.tokenize(["a diagram", "a dog", "a cat"])
+
+with torch.no_grad():
+    image_features = model.encode_image(image)
+    text_features = model.encode_text(text)
+    logits_per_image, logits_per_text = model(image, text)
```
[the remark this answers is not in the document]
Comment (lesters) [reply _posts/2021-11-28-text-image-search.md:106]: No, this is correct. The `image_features` and `text_features` are there to point out that images and text are transformed to representations, but that is also done inside `model`. As such, it is not necessary here, but I mention them in the text below.

File: _posts/2021-11-28-text-image-search.md (hunk 3 above)
Comment (lesters) [reply _posts/2021-11-28-text-image-search.md:248]: Changed it to `pyvespa` except when starting a sentence.

**Pull request merged**